tokio/src/runtime/task/trace/mod.rs

```diff
@@ -0,0 +1,215 @@
+    /// form a valid linked list.
+    unsafe fn with_current<F, R>(f: F) -> R
+    where
+        F: FnOnce(&Self) -> R,
+    {
+        crate::runtime::context::with_trace(f)
+    }
+
+    unsafe fn with_current_frame<F, R>(f: F) -> R
+    where
+        F: FnOnce(&Cell<Option<NonNull<Frame>>>) -> R,
+    {
+        Self::with_current(|context| f(&context.active_frame))
+    }
+
+    fn with_current_collector<F, R>(f: F) -> R
+    where
+        F: FnOnce(&Cell<Option<Trace>>) -> R,
+    {
+        unsafe { Self::with_current(|context| f(&context.collector)) }
+    }
+}
+
+impl Trace {
+    /// Invokes `f`, returning both its result and the collection of backtraces
+    /// captured at each sub-invocation of [`Trace::leaf`].
+    #[inline(never)]
+    pub(crate) fn capture<F, R>(f: F) -> (R, Trace)
+    where
+        F: FnOnce() -> R,
+    {
+        let collector = Trace { backtraces: vec![] };
+
+        let previous = Context::with_current_collector(|current| current.replace(Some(collector)));
+
+        let result = f();
+
+        let collector =
+            Context::with_current_collector(|current| current.replace(previous)).unwrap();
+
+        (result, collector)
+    }
+
+    /// The root of a trace.
+    #[inline(never)]
+    pub(crate) fn root<F>(future: F) -> Root<F> {
+        Root { future }
+    }
+
+    /// If this is a sub-invocation of [`Trace::capture`], capture a backtrace.
+    ///
+    /// The captured backtrace will be returned by [`Trace::capture`].
+    ///
+    /// Invoking this function does nothing when it is not a sub-invocation
+    /// [`Trace::capture`].
+    // This function is marked `#[inline(never)]` to ensure that it gets a distinct `Frame` in the
+    // backtrace, below which frames should not be included in the backtrace (since they reflect the
+    // internal implementation details of this crate).
+    #[inline(never)]
+    pub(crate) fn leaf() {
+        // Safety: We don't manipulate the current context's active frame.
+        unsafe {
+            Context::with_current(|context_cell| {
+                if let Some(mut collector) = context_cell.collector.take() {
+                    let mut frames = vec![];
+                    let mut above_leaf = false;
+
+                    if let Some(active_frame) = context_cell.active_frame.get() {
+                        let active_frame = active_frame.as_ref();
+
+                        backtrace::trace(|frame| {
+                            let below_root =
+                                !ptr::eq(frame.symbol_address(), active_frame.inner_addr);
+
+                            // only capture frames above `Trace::leaf` and below
+                            // `Trace::root`.
+                            if above_leaf && below_root {
+                                frames.push(frame.to_owned().into());
+                            }
+
+                            if ptr::eq(frame.symbol_address(), Self::leaf as *const _) {
```

Comment:
The tests are failing on MacOS because the unwinder never produces a symbol address exactly equalling `Self::leaf` — it's off by twelve bytes. I don't yet know why this occurs, or if the 12 byte offset is reasonably consistent or an artifact of codegen.

Reply:
I've opened an issue on backtrace-rs seeking further clarification: https://github.com/rust-lang/backtrace-rs/issues/520.

Reply:
The issue occurs because:
> On macOS the function to get the address of the enclosing function of an ip address (`_Unwind_FindEnclosingFunction`) is unreliable due to compact unwind info collapsing multiple functions with identical unwind info together

...and is thus not used. Consequently, unwinding on macOS with `backtrace-rs` currently always emits the same value for symbol address and instruction pointer.

There are a few ways we might work around this (or, even, fix the issue in `backtrace-rs`), but in the short term I think we should probably just drop macOS support for taskdumping from this initial PR.